tap_facebook/streams.py

```diff
@@ -193,6 +174,50 @@ def __get_start(self, account_id, state: dict, tap_stream_id: str) -> datetime:
         logger.info(f"using 'start_date' from previous state: {current_bookmark}")
         return new_date
 
+    def __run_adreport(
+        self,
+        account_id: str,
+        fields: Optional[Sequence[str]],
+        params: Dict[str, Any],
+    ) -> AdReportRun:
+        async_job = cast(
+            AdReportRun,
+            AdSet(account_id).get_insights(fields=fields, params=params, is_async=True),
+        )
+
+        while True:
+            job = cast(AdReportRun, async_job.api_get())
+
+            pct: int = job["async_percent_completion"]
+            status: str = job["async_status"]
+            job_id = job["id"]
+
+            if status in ["Job Skipped", "Job Failed"]:
+                logger.error(f"job<{job_id}>({params}): failed with status: {status}")
+                return async_job
+
+            # https://developers.facebook.com/docs/marketing-api/insights/best-practices/#asynchronous
+            # both fields need to be set to signify completion
+            if status == "Job Completed" and pct == 100:
+                job_start_time = datetime.fromtimestamp(job["time_ref"])
+                job_completion_time = datetime.fromtimestamp(job["time_completed"])
+                duration = job_completion_time - job_start_time
+
+                logger.info(f"job<{job_id}>: finished in {duration.seconds}s")
+
+                return async_job
+
+            if status in [
+                "Job Not Started",
+                "Job Started",
+                "Job Running",
+                "Job Completed",
```

Comment:
yes, edge case: job complete does not in fact mean that the job is done, it means that it is done fetching it, meaning we also have to wait for the pct == 100 as well. This is just to document that state, but I'll add a more explicit comment since this makes it clear that wasn't enough.